sympy/stats/tests/test_mix.py

```diff
@@ -18,6 +20,16 @@ def test_density():
     assert simplify(density(N2, Eq(N1, 1))(x)) == \
         sqrt(2)*exp(-(x - 1)**2/8)/(4*sqrt(pi))
 
+def test_MarginalDistribution():
+    a1, p1, p2 = symbols('a1 p1 p2', positive=True)
+    C = Multinomial('C', 2, p1, p2)
+    B = MultivariateBeta('B', a1, C[0])
+    MGR = MarginalDistribution(B, C[0])
+    mgrc1 = ("B*Piecewise((2*p1**C[0]*p2**C[1]/(factorial(C[0])*factorial"
+    "(C[1])), Eq(C[0] + C[1], 2)), (0, True))*gamma(a1 + C[0])*B[0]**(a1 - "
+    "1)*B[1]**(C[0] - 1)/(gamma(a1)*gamma(C[0]))")
+    assert str(MGR(C)) == mgrc1
```

Comment:
There are no dummies here...a non-string comparison should be ok.

Reply:
It doesn't works. I tried it. 

By the way can you please clear this,

> Is there any documentation which tells that how SymPy compares expressions? Like, I think knowing the concept will help me in using the right tools. Thanks.

Reply:
If you print the `srepr` of these objects you can see that they are very different:
```
In [20]: srepr(mgrc1)                                                                                                             
Out[20]: "Mul(Piecewise(ExprCondPair(Mul(Integer(2), Pow(Symbol('p1', positive=True), Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(0))), Pow(Symbol('p2', positive=True), Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(1))), Pow(factorial(Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(0))), Integer(-1)), Pow(factorial(Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(1))), Integer(-1))), Equality(Add(Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(0)), Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(1))), Integer(2))), ExprCondPair(Integer(0), true)), Pow(gamma(Symbol('a1', positive=True)), Integer(-1)), gamma(Add(Symbol('a1', positive=True), Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(0)))), Pow(gamma(Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(0))), Integer(-1)), Pow(Indexed(JointRandomSymbol(Symbol('B'), JointPSpace(Symbol('B'), MultivariateBetaDistribution(ImmutableDenseMatrix([[Symbol('a1', positive=True)], [Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(0))]])))), Integer(0)), Add(Symbol('a1', positive=True), Integer(-1))), Pow(Indexed(JointRandomSymbol(Symbol('B'), JointPSpace(Symbol('B'), MultivariateBetaDistribution(ImmutableDenseMatrix([[Symbol('a1', positive=True)], [Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(0))]])))), Integer(1)), Add(Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(0)), Integer(-1))), JointRandomSymbol(Symbol('B'), JointPSpace(Symbol('B'), MultivariateBetaDistribution(ImmutableDenseMatrix([[Symbol('a1', positive=True)], [Indexed(JointRandomSymbol(Symbol('C'), JointPSpace(Symbol('C'), MultinomialDistribution(Integer(2), ImmutableDenseMatrix([[Symbol('p1', positive=True)], [Symbol('p2', positive=True)]])))), Integer(0))]])))))"

In [21]: srepr(MGR(C))                                                                                                            
Out[21]: "Mul(Symbol('B'), Piecewise(ExprCondPair(Mul(Integer(2), Pow(Symbol('p1', positive=True), Indexed(IndexedBase(Symbol('C')), Integer(0))), Pow(Symbol('p2', positive=True), Indexed(IndexedBase(Symbol('C')), Integer(1))), Pow(factorial(Indexed(IndexedBase(Symbol('C')), Integer(0))), Integer(-1)), Pow(factorial(Indexed(IndexedBase(Symbol('C')), Integer(1))), Integer(-1))), Equality(Add(Indexed(IndexedBase(Symbol('C')), Integer(0)), Indexed(IndexedBase(Symbol('C')), Integer(1))), Integer(2))), ExprCondPair(Integer(0), true)), Pow(gamma(Symbol('a1', positive=True)), Integer(-1)), gamma(Add(Symbol('a1', positive=True), Indexed(IndexedBase(Symbol('C')), Integer(0)))), Pow(gamma(Indexed(IndexedBase(Symbol('C')), Integer(0))), Integer(-1)), Pow(Indexed(IndexedBase(Symbol('B')), Integer(0)), Add(Symbol('a1', positive=True), Integer(-1))), Pow(Indexed(IndexedBase(Symbol('B')), Integer(1)), Add(Indexed(IndexedBase(Symbol('C')), Integer(0)), Integer(-1))))"
```
That's why they don't compare equal with `==`.

Reply:
`srepr` -> sympy representation?

Reply:
> It doesn't works. I tried it.

Have you tried it after making things more canonical?

Reply:
It was not working previously because of the non-canonical Mul: sometimes the B came before the Piecewise. Now that you have that worked out and are getting only one string the comparision should no longer fail. The reason you thought you needed strings was because entering a SymPy expression as the test result *automatically* made it canonical, but SymPy routines can produce non-canonical output (but they shouldn't) and before your fix the MGR(C) was making non-canonical output.

MWE:
```python
def do(a, b):
  return Mul(b, a, evaluate=False)

>>> do(2, x)  # /!\ printing can rearrange args
2*x
>>> q = _
>>> q == 2*x  # no, q is actually x*2
False
>>> q == x*2  # no, typing 'x*2' becomes '2*x' so this is the same as before
False
>>> 2*x == x*2  # proof of previous
True
```

So how do you test it? Your answer was to use a string `str(do(2,x)) == 'x*2'` and that passes...but this could have been tested with `do(2, x) = Mul(x, 2, evaluate=False)`.  *But unless there is some good reason, output should not be non-canonical.*

Reply:
The problem is that the two things tested are really not the same even if they have similar str. The different position of `B` is due to the matrices that are in one expression but not the other so they canonicalise differently.

Reply:
LOL - the part that scrolled off the screen tells the story. I hadn't looked and based my remarks off of the two alternate forms that had been tested as strings. Thanks, @oscarbenjamin .

Reply:
This conversation was marked as resolved but I don't think it is resolved. You should work out how to create the object that is returned by MGR(C)

Reply:
> Have you tried it after making things more canonical?

Yes it fails.

May be what about,
```python
mgrc = Mul(Symbol('B'), Piecewise(ExprCondPair(Mul(Integer(2),
    Pow(Symbol('p1', positive=True), Indexed(IndexedBase(Symbol('C')),
    Integer(0))), Pow(Symbol('p2', positive=True),
    Indexed(IndexedBase(Symbol('C')), Integer(1))),
    Pow(factorial(Indexed(IndexedBase(Symbol('C')), Integer(0))), Integer(-1)),
    Pow(factorial(Indexed(IndexedBase(Symbol('C')), Integer(1))), Integer(-1))),
    Eq(Add(Indexed(IndexedBase(Symbol('C')), Integer(0)),
    Indexed(IndexedBase(Symbol('C')), Integer(1))), Integer(2))),
    ExprCondPair(Integer(0), True)), Pow(gamma(Symbol('a1', positive=True)),
    Integer(-1)), gamma(Add(Symbol('a1', positive=True),
    Indexed(IndexedBase(Symbol('C')), Integer(0)))),
    Pow(gamma(Indexed(IndexedBase(Symbol('C')), Integer(0))), Integer(-1)),
    Pow(Indexed(IndexedBase(Symbol('B')), Integer(0)),
    Add(Symbol('a1', positive=True), Integer(-1))),
    Pow(Indexed(IndexedBase(Symbol('B')), Integer(1)),
    Add(Indexed(IndexedBase(Symbol('C')), Integer(0)), Integer(-1))))
    assert MGR(C) == mgrc 
```
Though it looks too ad-hoc, but is there anything else we can do.

Reply:
You should translate that expression into normal SymPy code.